ipfix/test_ipfix.py

```diff
@@ -0,0 +1,135 @@
+
+from scapy.packet import Raw
+from scapy.layers.inet import IP, UDP, Ether
+from ipfix import IPFIX, Template
+
+from framework import VppTestCase, VppTestRunner
+from util import ppp
+from random import randint
+
+class TestIPFIX(VppTestCase):
+    """ IPFIX test case """ # test names are required
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestIPFIX, cls).setUpClass()
+        cls.create_pg_interfaces(range(2))  # create pg0 and pg1
+        for i in cls.pg_interfaces:
+            i.admin_up()  # put the interface upsrc_if
+            i.config_ip4()  # configure IPv4 address on the interface
+            i.resolve_arp()  # resolve ARP, so that we know VPP MAC
+
+    def setUp(self):
+        super(TestIPFIX, self).setUp()
+        # FIXME: I can't figure out how to get the test framework to call the
+        #        IPFIX plugin's API instead of using the CLI like this
+        self.logger.info(self.vapi.ppcli("set ipfix timeout template 1"))
+        self.logger.info(self.vapi.ppcli("set ipfix ip collector " + self.pg0.remote_ip4))
+        self.logger.info(self.vapi.ppcli("set ipfix ip exporter " + self.pg1.remote_ip4))
+        self.logger.info(self.vapi.ppcli("ipfix flow-meter " + self.pg1.name))
+
+    def test_template(self):
+        self.pg0.enable_capture()
+        self.pg1.enable_capture()
+        self.pg_start()
+        # no packets on pg1 since we didn't inject any packets
+        self.pg0.assert_nothing_captured()
+        # expect 1 template pkt
+        capture = self.pg0.get_capture(timeout = 1, expected_count = 1)
+        self.verify_template(self.pg0, self.pg1, capture)
+
+    # FIXME: this test doesn't work yet
+    #def test_basic(self):
+    #    packet_count = 10
+    #    packets = self.create_stream(self.pg0, self.pg1, packet_count)
+
+    #    self.pg0.add_stream(packets)
+    #    self.pg0.enable_capture()
+    #    self.pg1.enable_capture()
+    #    self.pg_start()
+
+    #    capture1 = self.pg1.get_capture(timeout = 1, expected_count = packet_count)
+    #    # with the timeout above, we expect to get a template pkt and one data pkt
+    #    capture0 = self.pg2.get_capture(timeout = 3, expected_count = 2, filter_out_fn=None)
+    #    self.verify_capture(self.pg0, self.pg1, capture0)
+
+    def verify_template(self, collector_if, exporter_if, capture):
+        for packet in capture:
+            try:
+                ip = packet[IP]
+                udp = packet[UDP]
+                ipfix = packet[IPFIX]
+                self.assert_equal(ip.src, exporter_if.remote_ip4,
+                                  "exporter ip")
+                self.assert_equal(ip.dst, collector_if.remote_ip4,
+                                  "collector ip")
+                self.assert_equal(udp.dport, 4739)
+                self.assert_equal(ipfix.version, 10)
+                # check that there's an IPFIX template
+                self.assert_equal(ipfix[Template].templateID, 256)
+            except:
```

Comment:
It's usually better to catch the specific errors you're looking for. Seems to me you're looking for both the key error (i.e. you can't find the part of the packet you expected to and thus it's thrown an error) and assertion errors (i.e. the packet was valid but not what you expected). I'm thinking something like this:
```python
try:
     ip = packet[IP]
except KeyError:
    # Don't know if this is the correct exception for packet[IPFIX] or others to throw exception
    self.logger.error(ppp("Invalid packet:", packet))
    raise
except AssertionError:
    self.logger.error(ppp("Unexpected packet:", packet))
    raise
```